### PR TITLE
Fix check_tsd_v2

### PR DESCRIPTION
**Pull request by bjozet**
This PR contains 3 commits. I've kept them separate for ease of reviewing, but they kind of all go together. 2 of the commits fixes broken behaviour, and 1 is cosmetic only.

See commit messages for individual commits to know more what they're doing.

### CODE REVIEW

**Comment by bjozet** [comment]
@johnseekins, you wrote the original file. Maybe you have some insights on this change, in case i've severely misunderstood something? Or perhaps just a review? :)

**Comment by johnseekins** [comment]
Happy to look in general. Can you give some insight into what you're fixing?

**Comment by bjozet** [comment]
@johnseekins, I saw your reply via mail only, it didn't seem to show up in this thread (perhaps github has some issues atm).

I'm fixing 2 issues:

1. The `-P` / `--percent-over` flag didn't work, becuase of flawed logic in lines 174-175](https://github.com/OpenTSDB/opentsdb/blob/master/tools/check_tsd_v2#L174-L175). It would raise a critical alarm regardless of parameter values. My commit c05f84b fixes that.
1. The `--ignore-recent` was also flawed and had inverse effect. If parameters `-d 180 -I 80` were used, only the latest 80s would be evaluated, and the 100 first were discarded. I also failed to see the benefit of query opentsdb for all data in duration`-d` and then discarding the datapoints up to `-I`, so i changed the parameters to modify the query-string to include a `end={}s-ago` parameter, so that opentsdb only sent the data we wanted.

If my description doesn't make sense, I'd be happy to explain it over a jitsi-meet or something.

**Comment by bjozet** [comment]
To clarify with examples:

This is running the original check, note how it finds timestamps "outside eval range" which clearly should be included with `-d 20 -I 3`, and also alerting despite setting `-P 50`:
```
08:59:15 user@host:~$ date +%s; ./check_tsd_v2 -H opentsdb.host.name -p 4242 -d 20 -I 3 -P 50 -m rate:tsd.rpc.received -t type=telnet -e lt,500,crit --debug
1586329157
Timestamp outside evaluation range: 1586329145.0
Timestamp outside evaluation range: 1586329141.0
Timestamp outside evaluation range: 1586329143.0
Timestamp outside evaluation range: 1586329138.0
Timestamp outside evaluation range: 1586329153.0
Timestamp outside evaluation range: 1586329150.0
Timestamp outside evaluation range: 1586329148.0
telnet => alarmed 1 times in range. (12.5% alarms). Avg. Value: 0.0
1 total metrics processed
1 Critical Alarms

08:59:17 user@host:~$ echo $?
2
```

This is my version, which handles the values correctly and does not alert, with the given percentage threshold (yes i had to skip the `-I 3` parameter in my test, otherwise i couldn't get it to alert):
```
09:20:19 user@host:~$ date +%s; ./check_tsd_v2_patch -H opentsdb.host.name  -p 4242 -d 20 -P 50 -m rate:tsd.rpc.received -t type=telnet -e lt,500,crit --debug
1586330421
Sending Query: http://opentsdb.host.name:4242/api/query?start=20s-ago&noAnnotations=true&m=sum:rate:tsd.rpc.received{type=telnet}
Processing timestamp 1586330406 value 10408.266666666666
Processing timestamp 1586330411 value 10299.5
Processing timestamp 1586330416 value 3311.1
Processing timestamp 1586330413 value 7004.6
Processing timestamp 1586330419 value 0.0
Processing timestamp 1586330410 value 10409.37777777778
Processing timestamp 1586330409 value 10408.266666666666
Datapoints outside of threshold: 1
Calculated Critical Percent: 14.285714285714285, less than value of percent_over argument: 50.0
1 total metrics processed

09:20:21 user@host:~$ echo $?
0
```

**Comment by johnseekins** [comment]
Yeah. Your logic all looks solid to me. I'd say maybe we talk about those two notes and then hope somebody with commit access comes across this, too.

File: tools/check_tsd_v2 [hunk 1]
```diff
@@ -149,35 +151,33 @@ def _process_metric(m, args, comparisons, now):
...
-                log.error("Bad timestamp for {}: {}".format(",".join(mresult["tags"]), ts))
-                mresult["crit_alarm"] = True
-                break
-            delta = now - ts
-            if delta >= args.ignore_recent:
-                log.debug("Timestamp outside evaluation range: {}".format(ts))
-                continue
+        log.debug("Processing timestamp {} value {}".format(ts,d))
+        try:
+            ts = float(ts)
```
[the remark this answers is not in the document]
Comment (bjozet) [reply tools/check_tsd_v2:156]: Well, yes. According to docs all timestamps are supposed to be interegers when using http-api, second/millisecond resolution is handled in a 10 vs 13 length unix-timestamp. I didn't change that line though, it was `float` in the original file also, i just changed indentation. However the whole check of timestamps is perhaps a bit redundant. Will opentsdb ever return with invalid timestamps? I'm guessing no, and furthermore, we don't ever use the "ts" variable anyway. Might as well remove the entire check, agree?

File: tools/check_tsd_v2 [hunk 2]
```diff
@@ -149,35 +151,33 @@ def _process_metric(m, args, comparisons, now):
...
     mresult["warn_percent"] = mresult["warn"] / value_count * 100
     if mresult["crit"] > 0:
-        if args.percent_over > 0 and mresult["crit_percent"] > args.percent_over:
-            mresult["crit_alarm"] = True
+        if args.percent_over > 0 and mresult["crit_percent"] < args.percent_over:
+            log.debug("Calculated Critical Percent: {}, less than value of percent_over argument: {}".format(mresult["crit_percent"], args.percent_over))
+            mresult["crit_alarm"] = False
         else:
             mresult["crit_alarm"] = True
     if mresult["warn"] > 0:
```
[the remark this answers is not in the document]
Comment (bjozet) [reply tools/check_tsd_v2:177]: I don't think that's necessary. It's already handled at L299-L302.

**Comment by johnseekins** [comment]
Awesome! Glad you caught all this. I probably would have, except we internally abandoned use of this tool right as I was submitting it.
@johann8384 @manolama This guy did some good work. Wanna merge it?

File: tools/check_tsd_v2 (hunk 1 above)
[the remark this answers is not in the document]
Comment (bjozet) [reply tools/check_tsd_v2]: pushed a commit that changed this to int(), even though i suspect the check is redundant.

**Comment by bjozet** [comment]
I also have question regarding the state of the `-A` flag, that's supposed to create a Critical Alert if no values were returned by opentsdb for a query, this is default to `false`. I suggest we make this default `true` instead, I cannot think of a scenario where you wouldn't wanna know if nothing was returned.

lines 243-244, thoughts? @johnseekins

**Comment by johnseekins** [comment]
Everything still looks good to me. Glad you've had some time to improve this tool.

**Pull request merged**